coordinator/src/services/translation/common/index.ts

```diff
@@ -173,14 +177,70 @@ export function convertMomentToHl7V3DateTime(dateTime: moment.Moment): core.Time
   return new core.Timestamp(hl7V3DateTimeStr)
 }
 
-export function wrapInOperationOutcome<T>(message: SpineDirectResponse<T>): fhir.OperationOutcome {
+function toArray<T>(thing: T | Array<T>): Array<T> {
+  return Array.isArray(thing) ? thing : [thing]
+}
+
+function translateSyncSpineResponse(message: string): Array<fhir.CodeableConcept> {
+  const parsedMsg = readXml(message) as SyncMCCI
+  const acknowledgementDetailElm = parsedMsg.MCCI_IN010000UK13.acknowledgement.acknowledgementDetail
+  const acknowledgementDetailArray = toArray(acknowledgementDetailElm)
+  return acknowledgementDetailArray.map(acknowledgementDetail => {
+    return {
+      coding: [{
+        code: acknowledgementDetail.code._attributes.code,
+        display: acknowledgementDetail.code._attributes.displayName,
+        system: ""
+      }]
+    }
+  })
+}
+
+function translateAsyncSpineResponse(message: string): Array<fhir.CodeableConcept> {
+  const parsedMsg = readXml(message) as AsyncMCCI
+  const reasonElm = parsedMsg["hl7:MCCI_IN010000UK13"]["hl7:ControlActEvent"]["hl7:reason"]
+  const reasonArray = toArray(reasonElm)
+  return reasonArray.map(reason => ({
+    coding: [{
+      code: reason["hl7:justifyingDetectedIssueEvent"]["hl7:code"]._attributes.code,
+      display: reason["hl7:justifyingDetectedIssueEvent"]["hl7:code"]._attributes.displayName,
+      system: ""
+    }]
+  }))
+}
+
+export function translateToOperationOutcome<T>(message: SpineDirectResponse<T>): fhir.OperationOutcome {
+  if (message.statusCode <= 299) {
+    return {
+      resourceType: "OperationOutcome",
+      issue: [{
+        code: "informational",
+        severity: "information",
+        diagnostics: message.body?.toString()
+      }]
+    }
+  }
+
+  const bodyString = message.body.toString()
+  const syncMCCI = SYNC_SPINE_RESPONSE_MCCI_REGEX.exec(bodyString)
+  const asyncMCCI = ASYNC_SPINE_RESPONSE_MCCI_REGEX.exec(bodyString)
+  let codeableConceptArray: Array<fhir.CodeableConcept> = []
+  if (syncMCCI) {
+    codeableConceptArray = translateSyncSpineResponse(syncMCCI[0])
+  } else if (asyncMCCI) {
+    codeableConceptArray = translateAsyncSpineResponse(asyncMCCI[0])
+  } else {
+    throw new Error("Spine response doesnt match")
```

Comment:
is this the best way to handle an unexpected spine error response?

Reply:
It's _a_ way :shrug:

Reply:
(I think it's fine as long as we log out the content of the response somewhere, which I think Axios already does for us)